File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.py
import sys
sys.setrecursionlimit(10 ** 6)
input = sys.stdin.readline

N, M, R = map(int,input().split())

graph = [[] for _ in range(N+1)]
visited = [0] * (N+1)
c = 1

for _ in range(M):
    a,b = map(int,input().split())
    graph[a].append(b)
    graph[b].append(a)

def dfs(v):
    global c
    visited[v] = c
    for i in graph[v]:
        if visited[i] == 0:
            c += 1
            dfs(i)

for i in range(N+1):
    graph[i].sort()

dfs(R)

for j in range(1, N+1):
    print(visited[j])

File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');
const [N, M, R] = input.shift().split(' ').map(Number);
const arr = input.map((v) => v.split(' ').map(Number));
const graph = [...Array(N + 1)].map(() => []);
const visited = Array(N).fill(0);
let cnt = 1;

// 무방향(양방향) 그래프 만들기
arr.map(([from, to]) => {
  graph[from].push(to);
  graph[to].push(from);
});

// 오름차순 정렬
graph.map((v) => v.sort((a, b) => a - b));

// DFS
const dfs = (node) => {
  // graph의 0번째 인덱스는 쓰지 않으므로 현재 노드번호에 -1의 방문여부 체크
  if (!visited[node - 1]) {
    // 방문 체크함과 동시에 해당 인덱스의 값은 1씩 증가하는 cnt로 넣어준다.
    visited[node - 1] = cnt;
    cnt++;
    for (const next of graph[node]) dfs(next); // 재귀
  }
};

dfs(R);

console.log(visited.join('\n'));